### DISCUSSION TITLE
Add libtextworker

### DISCUSSION THREAD

**Discussion by lebao3105** [Akascape/tkinter-toolkit#16] [Add this]
It's my own library made for my IDE, but that doesn't mean you can't use it (a new name should be here one day..)

It has:
* Tkinter things like a text edit widget with line numbers (not implemented by me), is-changed status, a custom about dialog, directory tree + list (on work) with auto refresh (initially)
* wxPython with widgets same as Tkinter
* Auto coloring for both toolkits, uses `darkdetect`
* JSON+INI configuration system and class for anything you like
* And other utility stuff

I wrote all myself, plus some copy-paste-modify codes so there are many bugs here (0.1.4 now has configuration system class being broken because of using both INI and JSON in one class). Debian packaging support is coming!

Get the source code here: https://gitlab.com/textworker/libtextworker_py

Documentation: https://lebao3105.github.io/libtextworker

Available on PyPi!

**Comment by lebao3105** [comment]
![Ảnh màn hình 2024-05-28 lúc 23 46 15](https://github.com/Akascape/tkinter-toolkit/assets/77564176/287e6fbf-edd5-4618-bd73-115f39548721)
![Ảnh màn hình 2024-05-28 lúc 23 46 33](https://github.com/Akascape/tkinter-toolkit/assets/77564176/ab375aa0-3a2b-4d22-acaa-53a4b78effa4)

Here are some screenshots of the toolkit

**Reply by Akascape** [reply]
@lebao3105 I will add this, but why there is no github repository for this?

**Reply by lebao3105** [reply]
> @lebao3105 I will add this, but why there is no github repository for this?

In fact it has in my profile, but I moved+created some of my projects to GitLab instead of GitHub. So now the one on GitHub is archived. 

If that's required, I will make a mirror to GitHub.

**Reply by Akascape** [reply]
@lebao3105 It's better to make a repo on github, so that we can find the source code easily.

**Reply by lebao3105** [reply]
I created a mirror here: https://github.com/lebao3105/libtextworker

**Reply by Akascape** [reply]
@lebao3105 How helpful this project is for tkinter?

> Tkinter things like a text edit widget with line numbers (not implemented by me)

We already have tkLineNumbers for this, but ok 

> a custom about dialog, directory tree + list (on work) with auto refresh (initially)

Can you show some examples/screenshots of those widgets? 

> wxPython with widgets same as Tkinter

It's related to wxPython, not tkinter, so how will it help in tkinter projects? 

> Auto coloring for both toolkits, uses darkdetect
JSON+INI configuration system and class for anything you like

Does it work in tkinter? Can we save color properties to a json/ini file and load it again?

> And other utility stuff

Please mention other utility that are helpful for tkinter.

**Comment by lebao3105** [comment]
> We already have tkLineNumbers for this, but ok

The editor... uses tkLineNumbers, not strictly required and is used for an easier setup (turn on/off via settings file)

> It's related to wxPython, not tkinter, so how will it help in tkinter projects?

None. My description goes too far. Sorry about that.

> Does it work in tkinter? Can we save color properties to a json/ini file and load it again?

Yes for both two questions.

> Please mention other utility that are helpful for tkinter.

These 'utilities' that I wanted to mention are even not GUI-related - they are something like comparing between version. However Tkinter has a function to create a menu with items inside (no sub-menu), it should have functions about events binding and more.

> Can you show some examples/screenshots of those widgets?

Here goes the about dialog:

![Ảnh màn hình 2024-05-29 lúc 21 58 16](https://github.com/Akascape/tkinter-toolkit/assets/77564176/dd31b1f6-9f97-44f9-af0f-94a09d853cae)

![Ảnh màn hình 2024-05-29 lúc 21 58 35](https://github.com/Akascape/tkinter-toolkit/assets/77564176/f551af8e-9652-42fa-aefd-0e4462062263)

![Ảnh màn hình 2024-05-29 lúc 21 58 58](https://github.com/Akascape/tkinter-toolkit/assets/77564176/a9f11a21-42f1-4a09-8cf8-968b93fd89c0)

```python
aboutdlg = AboutDialog()
aboutdlg.SetProjectName("libtextworker")
aboutdlg.SetProjectVersion(libver)
aboutdlg.SetProjectDescription("A Python library made for GUI apps.")
aboutdlg.SetProjectSite(GITHUB_URL)
aboutdlg.SetAppTesters(["Le Bao Nguyen"])
aboutdlg.SetDevelopers(["Le bao Nguyen"])
aboutdlg.SetProjectLicense("GNU General Public License Version 3")
```

The directory tree's auto refresh feature currently doesn't work (Tkinter widget's .event_generate() does not allow custom keyword for the file system event path)

**Comment by lebao3105** [comment]
Here's how the file system event should work: it pushes a new virtual event, we just bind the event to a logging widget (auto refreshing isn't available now) or show a message box (what the Tkinter test does)

https://github.com/Akascape/tkinter-toolkit/assets/77564176/dedc8da8-564b-4e6b-9492-449c90927411

**Comment by Akascape** [comment]
@lebao3105 Thanks for giving all the details, I have added it in the database.
